### ISSUE TITLE
Our iOS perf is really bad or 'Make sure docs replicate' fails on iOS devices in WiFi mode

### ISSUE THREAD

**Issue by chapko** [thaliproject/Thali_CordovaPlugin#1618]
*'Make sure docs replicate'* fails on iOS devices with the following error:

```
# Make sure docs replicate
2016-12-07 15:20:50 - DEBUG thaliReplicationPeerAction: 'Replication resumed'
2016-12-07 15:20:54 - DEBUG localSeqManager: 'Got an error trying to update seq {"code":"ECONNRESET","status":500}'
2016-12-07 15:20:54 - DEBUG thaliReplicationPeerAction: 'Got error on replication -  500 socket hang up -1'
2016-12-07 15:20:54 - DEBUG thaliReplicationPeerAction: 'We called _complete with error: socket hang up -1  and it caused us to complete'
2016-12-07 15:20:54 - DEBUG thaliReplicationPeerAction: 'Got error in update, waiting for main loop to detect and handle -  500 socket hang up -1'
not ok 753 failed with Error: Could establish TCP connection but couldn't keep it running
  ---
    operator: fail
  ...
2016-12-07 15:20:54 - ERROR CoordinatedClient: 'test failed, name: 'Make sure docs replicate''
```

When I run this test file (testThaliReplicationPeerAction.js) first it always passes.

**Comment by chapko** [comment]
Device logs:
- [2.log](https://gist.github.com/chapko/b87f3f3ffc1bfa4fef8c9f578a9b2676/raw/f326dca8a050722dd5b5f158063a07cb1b4760a0/2.log)
- [5.log](https://gist.github.com/chapko/b87f3f3ffc1bfa4fef8c9f578a9b2676/raw/f326dca8a050722dd5b5f158063a07cb1b4760a0/5.log)
- [7.log](https://gist.github.com/chapko/b87f3f3ffc1bfa4fef8c9f578a9b2676/raw/f326dca8a050722dd5b5f158063a07cb1b4760a0/7.log)

**Comment by chapko** [comment]
Turns out jxcore on iOS devices is really slow. I have put a duplicate of 'Make sure docs replicate' before other tests, so it runs twice: the first one at the beginning and the second one near the end. In the first replication test updating sequence number on the "remote" DB (in this test both local and remote databases are on the same device)  takes 800-900 milliseconds. The second one - around 3400 milliseconds. But we have only 3 seconds no-activity timeout after which action is killed, localSeqManager is stopped and seq number update is rejected.

One of the problems here is that there is no way to observe sequence updating within replication action because it is not exposed.

I think sequence number updating process should be treated as an activity by "No activity" timer.

Also I noticed one thing in sequence updating process (`_doImmediateSeqUpdate`):
1. We are making get request to get our revision from remote db (`getLocalIdRevOrNull`).
2. We are making put request to put doc containing our `_localId`, `seq` number and `rev` from the 1st step
3. We update `_secDocRev` value with the `rev` property of the put response from the 2nd step.

It looks like the first step is pointless and just takes extra time (~1.2 seconds of the 3.4 total):
```
2016-12-08 10:41:53.660 - INFO localSeqManager::upd #13: 'runUpdate: waiting for blocked request'
2016-12-08 10:41:53.665 - INFO localSeqManager::imm #13: 'Requesting remote db for local id: _local/thali_BKhVe6NKGpWU9QXkvveVD8vi-xxScwV12XgnghwI9zTMF489i4fnY2yF3z7ScIB9ZkC69hpRB_BUDUHAB6yctkA'
2016-12-08 10:41:54.840 - INFO localSeqManager::imm #13: 'Putting doc {"_id":"_local/thali_BKhVe6NKGpWU9QXkvveVD8vi-xxScwV12XgnghwI9zTMF489i4fnY2yF3z7ScIB9ZkC69hpRB_BUDUHAB6yctkA","_rev":null,"lastSyncedSequenceNumber":10} to the remote db'
2016-12-08 10:41:57.098 - INFO localSeqManager::imm #13: 'Got put response: {"ok":true,"id":"_local/thali_BKhVe6NKGpWU9QXkvveVD8vi-xxScwV12XgnghwI9zTMF489i4fnY2yF3z7ScIB9ZkC69hpRB_BUDUHAB6yctkA","rev":"0-1"}'
2016-12-08 10:41:57.099 - INFO localSeqManager::upd #13: 'runUpdate: completed with "0-1"'
SEQ UPDATE TOOK 3447 MILLISECONDS
```

**Comment by yaronyg** [comment]
So @chapko please create a trivial little demo (feel free to use thaliTestApp) where you do a get followed by a post with no PouchDB and see how long it takes. If it's super fast then please add in PouchDB so it's a doc get and a doc update. If that is super fast then we have to assume that this has to do with running a replication at the same time as the separate PouchDB get and update calls and investigate that.

**Comment by yaronyg** [comment]
Tests we should run:

1. Run 50 HTTP GET requests in a row, store each one's time from issuing the request to getting the response and then print the average, median and standard deviation.
2. Run 5 HTTP GET requests in parallel and for each one store each one's time form issuing the request to getting the response and then print the average, median and standard deviation.
3. Repeat test 2 for 10 and 15 HTTP GET requests in parallel.

**Comment by squid48** [comment]
@yaronyg here is the result:
#### Call GET 50 times in row mode
100ms, 111ms, 121ms, 129ms, 127ms, 130ms, 129ms, 130ms, 128ms, 129ms, 129ms, 129ms, 130ms, 129ms, 128ms, 131ms, 129ms, 129ms, 130ms, 129ms, 129ms, 130ms, 129ms, 129ms, 130ms, 130ms, 128ms, 135ms, 141ms, 130ms, 129ms, 129ms, 129ms, 133ms, 131ms, 139ms, 131ms, 129ms, 130ms, 129ms, 129ms, 129ms, 130ms, 129ms, 129ms, 130ms, 129ms, 129ms, 129ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
128.77551020408163ms | 129ms |  5.628880724360712ms

#### Call GET 5 times in parallel mode
194ms, 192ms, 191ms, 189ms, 189ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
191ms | 191ms | 1.8973665961010275ms

#### Call GET 10 times in parallel mode
194ms, 194ms, 194ms, 194ms, 194ms, 194ms, 193ms, 193ms, 192ms, 192ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
193.4ms | 194ms | 0.7999999999999999ms

#### Call GET 15 times in parallel mode
213ms, 213ms, 213ms, 213ms, 213ms, 213ms, 213ms, 213ms, 212ms, 212ms, 213ms, 213ms, 213ms, 212ms, 213ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
212.8ms | 213ms | 0.4ms

**Comment by yaronyg** [comment]
@squid48 Can you please publish here the same tests but running on desktop so we can compare?

Now we need to test with PouchDB. Most of our requests are GET requests (because that is what replication does) so let's re-run the same tests as above but instead of a HTTP GET run a document GET on the PouchDB client. You can always ask for the same document ID in all the tests. It will be served out of RAM. That is useful because it lets us know if the problem is PouchDB's node code or maybe something with leveldown. Just to be clear, I'm specifying that we are reading exactly one document over and over and over. That way we are just testing PouchDB's perf without worrying about Leveldown-Mobile.

If those numbers aren't bad enough to explain the problem then please run the same tests as above but instead of a document GET, do a document PUT for a randomly generated ID. Let's see how bad the write perf is.

**Comment by squid48** [comment]
@yaronyg ok, here is the result on desktop
@yaronyg here is the result:
#### Call GET 50 times in row mode
25ms, 6ms, 4ms, 3ms, 5ms, 4ms, 4ms, 3ms, 5ms, 4ms, 3ms, 3ms, 3ms, 4ms, 5ms, 2ms, 4ms, 4ms, 4ms, 3ms, 2ms, 4ms, 3ms, 3ms, 3ms, 3ms, 4ms, 3ms, 3ms, 4ms, 7ms, 3ms, 4ms, 2ms, 2ms, 2ms, 3ms, 3ms, 4ms, 3ms, 3ms, 3ms, 3ms, 4ms, 3ms, 3ms, 3ms, 3ms, 5ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
 3.938775510204082ms | 3ms |  3.196989420965855ms

#### Call GET 5 times in parallel mode
13ms, 13ms, 13ms, 13ms, 14ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
13.2ms | 13ms |  0.4ms

#### Call GET 10 times in parallel mode
34ms, 33ms, 33ms, 34ms, 33ms, 34ms, 34ms, 34ms, 32ms, 32ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
33.3ms | 33.5ms | 0.7810249675906654ms

#### Call GET 15 times in parallel mode
43ms, 44ms, 44ms, 43ms, 43ms, 42ms, 43ms, 44ms, 43ms, 42ms, 41ms, 43ms, 43ms, 42ms, 42ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
42.8ms | 43ms | 0.8326663997864531ms

**Comment by squid48** [comment]
Table for a better comparison of the desktop and mobile results.

#### Call GET 50 times in row mode

 &nbsp; | Desktop | Mobile
------------ | ------------- | -------------
**Average** | 3.938775510204082ms |  128.77551020408163ms
**Median** | 3ms | 129ms
**Standard deviation** | 3.196989420965855ms | 5.628880724360712ms

#### Call GET 5 times in parallel mode

 &nbsp; | Desktop | Mobile
------------ | ------------- | -------------
**Average** | 13.2ms |  191ms
**Median** | 13ms | 191ms
**Standard deviation** | 0.4ms | 1.8973665961010275ms

#### Call GET 10 times in parallel mode

 &nbsp; | Desktop | Mobile
------------ | ------------- | -------------
**Average** | 33.3ms |  193.4ms
**Median** | 33.5ms | 194ms
**Standard deviation** | 0.7810249675906654ms | 0.7999999999999999ms

#### Call GET 15 times in parallel mode

 &nbsp; | Desktop | Mobile
------------ | ------------- | -------------
**Average** | 42.8ms |  212.8ms
**Median** | 43ms | 213ms
**Standard deviation** | 0.8326663997864531ms | 0.4ms

**Comment by yaronyg** [comment]
@squid48 The results are weird. I compared the medians and for GET 50 times in a row Mobile is 43x slower but for GET 15 times in parallel mode Mobile is only 5x slower. Weird.

But now we need to run the PouchDB test where we do a PouchDB.get request on a local HTTP URL to the local Express server and get the same doc. E.g. the same tests as above but getting the same PouchDB doc instead of just a direct HTTP GET request.

**Comment by squid48** [comment]
@yaronyg ok, here is the result on desktop

#### Call GET 50 times in row mode
25ms, 6ms, 4ms, 3ms, 5ms, 4ms, 4ms, 3ms, 5ms, 4ms, 3ms, 3ms, 3ms, 4ms, 5ms, 2ms, 4ms, 4ms, 4ms, 3ms, 2ms, 4ms, 3ms, 3ms, 3ms, 3ms, 4ms, 3ms, 3ms, 4ms, 7ms, 3ms, 4ms, 2ms, 2ms, 2ms, 3ms, 3ms, 4ms, 3ms, 3ms, 3ms, 3ms, 4ms, 3ms, 3ms, 3ms, 3ms, 5ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
 3.938775510204082ms | 3ms |  3.196989420965855ms

#### Call GET 5 times in parallel mode
13ms, 13ms, 13ms, 13ms, 14ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
13.2ms | 13ms |  0.4ms

#### Call GET 10 times in parallel mode
34ms, 33ms, 33ms, 34ms, 33ms, 34ms, 34ms, 34ms, 32ms, 32ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
33.3ms | 33.5ms | 0.7810249675906654ms

#### Call GET 15 times in parallel mode
43ms, 44ms, 44ms, 43ms, 43ms, 42ms, 43ms, 44ms, 43ms, 42ms, 41ms, 43ms, 43ms, 42ms, 42ms

Average time | Median | Standard deviation
------------ | ------------- | -------------
42.8ms | 43ms | 0.8326663997864531ms

Call makeGet 50 times in row mode
6ms, 3ms, 4ms, 5ms, 4ms, 4ms, 4ms, 6ms, 5ms, 5ms, 5ms, 5ms, 4ms, 6ms, 6ms, 4ms, 5ms, 3ms, 4ms, 5ms, 5ms, 16ms, 5ms, 6ms, 4ms, 5ms, 5ms, 5ms, 5ms, 13ms, 7ms, 3ms, 4ms, 4ms, 5ms, 3ms, 4ms, 3ms, 3ms, 4ms, 5ms, 5ms, 3ms, 4ms, 3ms, 3ms, 5ms, 3ms, 3ms
Average time: 4.816326530612245ms
Median: 5ms
Standard deviation: 2.2558175981747506ms
Call makeGet 5 times in parallel mode
9ms, 11ms, 13ms, 14ms, 15ms
Average time: 12.4ms
Median: 13ms
Standard deviation: 2.1540659228538015ms
Call makeGet 10 times in parallel mode
17ms, 21ms, 21ms, 23ms, 25ms, 30ms, 33ms, 39ms, 41ms, 42ms
Average time: 29.2ms
Median: 27.5ms
Standard deviation: 8.681013765684282ms
Call makeGet 15 times in parallel mode
14ms, 17ms, 18ms, 19ms, 20ms, 23ms, 25ms, 26ms, 28ms, 29ms, 31ms, 32ms, 35ms, 38ms, 39ms
Average time: 26.266666666666666ms
Median: 26ms
Standard deviation: 7.531858262665921ms

**Comment by squid48** [comment]
&nbsp; | 5 GET Parallel | 4 GET + 1 POST
------------ | ------------- | -------------
**Average** | 343ms |  **GET** 352.75ms; **POST** 494ms
**Median** | 344ms | **GET** 352.5ms; **POST** 494ms
**Standard deviation** | 70.85760368513742ms | **GET** 64.34817402226733ms; **POST** 0ms

 &nbsp; | 10 GET Parallel | 7 GET + 3 POST
------------ | ------------- | -------------
**Average** | 772.9ms |  **GET** 638.8571428571429ms; **POST** 1075.6666666666667ms
**Median** | 734ms | **GET** 604ms; **POST** 1093ms
**Standard deviation** | 227.4169958468364ms | **GET** 143.13173609493813ms; **POST** 61.65134944905009ms

 &nbsp; | 15 GET Parallel | 9 GET + 6 POST
------------ | ------------- | -------------
**Average** | 1056.4ms |  **GET** 851.2222222222222ms; **POST** 1543.8333333333333ms
**Median** | 1047ms | **GET** 778ms; **POST** 1536ms
**Standard deviation** | 366.2667152044623ms | **GET** 223.09777516584654ms; **POST** 179.84290984696122ms

**Comment by yaronyg** [comment]
@squid48 Also please link to the actual test code for the tests so we can look at that and make sure the tests are doing what we think they are doing.

**Comment by squid48** [comment]
@yaronyg I sent an invitation to the repository for you.

These tests were run without PSK and, as you can see, results are bad.
I have done all of the above changes and the result remained the same.

**Comment by yaronyg** [comment]
@squid48 Please make the testPouchDB and testDelay repositories public and put in the links to them here so that we have them as reference for the future.

O.k. so what we can see from the results is that PouchDB isn't really operating in parallel. It's operating serially. So now the question is - why?

If we think of the processing path there are several steps:

Client ---> Wire ---> Express ---> Express-PouchDB ----> PouchDB ---> LevelDown-Mobile

What we eventually have to do is narrow this down so we know which on it is. 

The tests we would run are:

Client - Run the TestPouchDB test but point it at a CouchDB server and see the results.

Local + MemDown - Re-write the test to not use the HTTP interface at all but instead use the local PouchDB server APis to do get and put/post and run against memdown instead of leveldown-mobile.

If we see a problem with client then we know that PouchDB isn't making requests properly in parallel.

If we see a problem with Local + MemDown then we know that PouchDB can handle requests in parallel but either MemDown or Express-PouchDB are screwing things up.

It's all about triangulation.

But for right now we aren't going to worry about it. I think we should first get all tests passing (including increasing the time outs for the seq tests on iOS) and then come back to this bug.

So again, @squid48, please make the two repo's you used public and please link them to this bug and then assign the bug to me. Thanks!

**Comment by squid48** [comment]
@yaronyg 
https://github.com/squid48/TestDelay
https://github.com/squid48/TestPouchDB

**Comment by chapko** [comment]
I just measured every single request time while running tests. Results are pretty spectacular: https://github.com/chapko/thali-logs/tree/1618. At the end of the tests some requests take up to 16 seconds.

I measured a time between request creation and emitting 'response' event. Code I used:
```JavaScript
var http = require('http');
var ClientRequest = http.ClientRequest;

// One of the first things ClientRequest constructor does is `this.agent = options.agent`
Object.defineProperty(ClientRequest.prototype, 'agent', {
  get: function () {
    return this.___agent;
  },
  set: function (v) {
    var self = this;
    if (!this.___agentSet) {
      this.___agentSet = true;
      this.___start = Date.now();
      this.once('response', function () {
        console.log('=== Request time: %sms (%s)', Date.now() - self.___start, self.path);
      });
      this.once('error', function () {
        console.log('=== Request error after %sms (%s)', Date.now() - self.___start, self.path);
      })
    }
    this.___agent = v;
  }
});
```

**Comment by yaronyg** [comment]
In a perfect world we would run the same test on Android just to see if this is a JXcore issue or an iOS issue.

**Comment by enricogior** [comment]
@yaronyg I ran again the test https://github.com/squid48/TestDelay to understand the weird results.
I ran it on desktop, iOS, Android with JIT enabled and Android with JIT disabled.

Here are my findings:
 - desktop and Android results are consistent, therefore we can just compare Android vs iOS.
 - JIT is not a factor, on Android a single https round-trip takes ~25ms when JIT is enable, it takes ~30ms when JIT is disabled, therefore it's consistent with the 20% perf impact we already knew and cannot be the main cause of the 130ms roundtrip occurring on iOS and the overall weirdness of the test results.
 - I ran both https and http tests, the results are consistent, I won't go into details here, beside pointing out  that https is on average ~250% slower than http (https->http: Android 25ms->9ms, iOS 130ms->54ms).
 - I also run a pure TCP test, because I wanted to reduce the test to the most simplified case possible. The TCP results are consistent with the http(s) results minus the overhead of http, therefore we can just analyze those (http->tcp: Android 9ms->4ms, iOS 54ms->30ms).

The client and the server are running in the same process and therefore with zero network latency.
On **Android**, a client-server TCP round-trip takes at max 4ms (min 2ms, but let's assume 4ms to simplify).
I ran multiple client requests in parallel starting each request at regular intervals.
If the interval between requests is over 4ms, each request will be atomic (i.e. nagling will not occur).
If the interval is 4ms or less, nagling is occurring almost all the times.
In some cases if it doesn't occur on the client, it will occur on the server.

On **iOS**, each individual request takes from 28 to 30ms.
Parallel requests trigger nagling on both the client and the server when the interval is less than 30ms. For example two requests 10ms apart, will trigger nagling on the client, if they are 25ms apart the nagling will occur on the server when the server is replying to the two requests that has received individually.

That explains the 'weird' results we were seeing in the TestDelay test. They are consistent with my results:

| Test | Android | iOS | x slower  |
|--|--|--|--|
|1 req| 4ms| 30ms | 7.5x |
|3 req| 14ms| 30ms | 2.14x |
|5 req 5ms interval (0-20)| 24ms| 30ms| 1.25x |
|10req 5ms interval (0-45) | 49ms | 57ms | 1.16x |

Next step was to test `socket.setNoDelay(true|false)`: unfortunately it doesn't seems to have any impact on iOS or Android. There may be two reasons why it doesn't work:
 1 - libuv doesn't set the TCP_NODELAY flag correctly
 2 - the OS ignores the flag.

I suspect the second, but I still need to complete my investigation to verify it.

**Comment by enricogior** [comment]
All the tests run so far, were executed with client and server on the same device and with the external device connectivity off (wifi, bt, radio).
Repeating the tests using a server running on the desktop and having all connectivity on shows unpredictable results.
Nagling may occur or not, randomly and up to 100ms (that is the higher value observed, but it may be even higher as far as we know).
Clearly, iOS is using its policies without allowing the application to force the TCP_NODELAY option.

Next step will be the investigation of chapko's findings, since 16s requests are not caused by the OS behavior.

**Comment by chapko** [comment]
@enricogior please, keep in mind that these 16s request in my last comments are long-polling requests and they are expected to be long when the server has no updates. The question is why server takes so long to reply and whether it is pouchdb, leveldown, jxcore or something else. The point of my measurements was that the same test has significantly different timing at the beginning of the tests run and at the end of tests run.

**Comment by enricogior** [comment]
After testing the basic client process scenario using TCP, HTTP and HTTPS:
```
Client ---> Wire --->Server 
```
we compared:
```
Client ---> Wire ---> Express ---> Express-PouchDB ----> PouchDB ---> LevelDown-Mobile
vs
Client ---> Wire ---> Express ---> Express-PouchDB ----> PouchDB ---> MemDown
```
Keep in mind that `LevelDown-Mobile` is implemented in C++, while `MemDown` is pure JS code, so even if `MemDown` is in-memory only, it may be slower.
Both on Android and iOS, the impact of switching to `MemDown` is greater than expected, in particular on iOS, queries are 100% slower, that doesn't sounds right, since it's a simple read from an in-memory balanced binary tree and it's only a small amount of code compared to the full client-server round-trip.

Upon further investigation it turned out that the overall CPU time used by the test is almost identical when using `LevelDown-Mobile` or `MemDown`, the real difference is in the average CPU usage that is higher when using `LevelDown-Mobile` (~20%) and lower when using `MemDown` (~12%). That may be caused by the libuv loop scheduler. We will continue the investigation testing the libuv loop behavior on both iOS and Android to identify if it is the cause of the performance difference.

**Comment by enricogior** [comment]
After further investigation, it turned out the performance issue was not caused by the fact that JXcore doesn't use JIT on iOS or by any other JXcore internal issue on iOS.
The problem was caused by how JXcore-Cordova interacts with the JXcore library.
JXcore-Cordova needs to control the libuv's loop because it needs to be able to run the native layer methods and callbacks upon request without having to wait for libuv if there are async I/O operations pending.
The way, both Android and iOS manage that mechanism, is similar, but on iOS it had higher values for the delay between libuv's loop executions.
In short, every async I/O operation had a 10ms delay that sums up with other async operations so the result was the higher execution times reported by the tests. On Android the delay is 1ms and has a lower impact.
Given that Thali components are doing very little CPU intensive operations and a lot of async I/O, the performance impact on iOS was huge.

We re-ran the tests https://github.com/squid48/TestPouchDB before and after the fix:
```
BEFORE:
PUT: 750ms, 189ms, 184ms, 185ms, 183ms, 184ms, 181ms, 184ms, 186ms, 180ms, 184ms, 184ms, 184ms, 185ms, 182ms, 181ms, 186ms, 189ms, 182ms, 186ms, 191ms, 184ms, 184ms, 182ms, 330ms, 188ms, 182ms, 185ms, 184ms

AFTER:
PUT: 206ms, 9ms, 8ms, 8ms, 8ms, 8ms, 10ms, 9ms, 8ms, 8ms, 7ms, 9ms, 7ms, 9ms, 8ms, 9ms, 7ms, 9ms, 8ms, 17ms, 14ms, 10ms, 8ms, 9ms, 8ms, 8ms, 8ms, 8ms, 8ms

---------------------------------------------
BEFORE:
Call makeGet 4 times in parallel mode
Call makePost 1 times in parallel mode
GET: 264ms, 318ms, 369ms, 421ms
POST: 467ms

AFTER:
Call makeGet 4 times in parallel mode
Call makePost 1 times in parallel mode
GET: 16ms, 20ms, 24ms, 27ms
POST: 27ms

---------------------------------------------
BEFORE:
Call makeGet 7 times in parallel mode
Call makePost 3 times in parallel mode
GET: 462ms, 514ms, 566ms, 621ms, 676ms, 728ms, 780ms
POST: 830ms, 888ms, 935ms

AFTER:
Call makeGet 7 times in parallel mode
Call makePost 3 times in parallel mode
GET: 31ms, 33ms, 37ms, 40ms, 44ms, 47ms, 54ms
POST: 56ms, 61ms, 61ms

---------------------------------------------
BEFORE:
Call makeGet 9 times in parallel mode
Call makePost 6 times in parallel mode
GET: 785ms, 837ms, 891ms, 945ms, 1001ms, 1052ms, 1104ms, 1153ms, 1217ms
POST: 1273ms, 1333ms, 1390ms, 1449ms, 1501ms, 1547ms

AFTER:
Call makeGet 9 times in parallel mode
Call makePost 6 times in parallel mode
GET: 46ms, 49ms, 53ms, 56ms, 60ms, 64ms, 67ms, 71ms, 75ms
POST: 81ms, 83ms, 88ms, 92ms, 97ms, 98ms
```
In those tests we see improvements between 10X and 25X.
In other tests we found improvements up to 100X.
The fix makes Thali and JXcore on iOS run way faster than on Android.
We haven't fixed Android yet, we will test a similar fix on it as well.

**Comment by enricogior** [comment]
We also verified the CPU usage and the battery life impact, so far things look good with low battery life impact reported by the Xcode tool. We will have to do more tests but we expect actually an improvement in real life situation given that now data synchronization should take 1/10th of the time or less, that in turns means less time while the app need to stay in foreground to complete the data transfers.

**Comment by enricogior** [comment]
We opted to leave the delay, but lowering it to 1ms instead of completely remove it.
The reason is that without the delay, the calling code needs to be aware that long timeouts may cause high cpu usage. In order to minimize the impact to the existing code it's a safer option, at least for the time being, to use the delay at 1ms similarly to the Android implementation.
In the future we can revisit this issue.

EDIT: 1ms seems too aggressive, increasing the CPU usage when the app is idle, 5ms seems a better option to have a balance between better performance and low battery usage.
